Enhancement/logging datetime

Guilty-Spark-343 opened the pull request:
### Prerequisites

- [X] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change

As outlined in #1058, the logs don't include the datetime which can make debugging difficult.  The proposed change adds a new formatter and attaches it to every handler when the application starts.
![image](https://github.com/constellation-app/constellation/assets/63685177/590e4ebe-08ca-4ef2-be20-787cb95ddf2c)


### Alternate Designs

No alternate designs considered.

### Why Should This Be In Core?

This is a change to the current logging in the core and provides a consistent logging format.

### Benefits

Consistent logging format that includes the datetime that makes debugging easier.

### Possible Drawbacks

None considered.

### Verification Process

1. Open Constellation
2. Click Help in the top banner, then 'Show Logs'
3. Obverse the datetime string at the start of each log e.g. [2023-12-28 10:26:38.81Z]

Further testing can be done via a Scripting View script:
```
print('\nLOG TEST SCRIPT STARTED')

from time import sleep
from java.lang import Runnable
from java.lang import Thread
from java.util.logging import Logger
from java.util.logging import Level
import java.lang.Exception
import java.lang.Throwable

LOGGER = Logger.getLogger("ScriptingView instance");
LOGGER.setLevel(Level.ALL);

class FineErrorThread(Runnable):
    def run(self):
        try:
            raise java.lang.Exception("Generated exception: type FINE")
        except java.lang.Throwable, t:
            LOGGER.log(Level.FINE, "Re-routed exception ", t);
        
LOGGER.log(Level.FINE, "\n\n  << Starting LOG test >>")
LOGGER.log(Level.INFO, " << Generating 25 message logs ... >>")
for idx in range(0, 25):
    LOGGER.log(Level.INFO, " ** This message will be logged multiple times **")
    sleep(0.05)  

LOGGER.log(Level.WARNING, " << Generating 15 Exception logs ... >>")
for idx in range(0, 15):
    testseverrorThread = Thread(FineErrorThread())
    testseverrorThread.start()
    sleep(0.05)  

sleep(0.75) 
LOGGER.log(Level.SEVERE, " << Completed LOG test >>")
print('\nLOG TEST SCRIPT COMPLETED')

```

### Applicable Issues

#1058

antares1470 reviewed (changes requested):
Thanks for doing this @Guilty-Spark-343.
Just looking at it in testing, I'm wondering if its possible to retain the logger name (to track where the log came from) and log level as these are removed with the new formatter.

Guilty-Spark-343 commented:
Good idea @antares1470! I've added the logger level and logger name.  It's in the format:
 [<datetime>] - <Logging level> <logger name>: <message>
![image](https://github.com/constellation-app/constellation/assets/63685177/bb38b566-047b-4560-8418-5635f8a8d21c)

antares1470 reviewed (changes requested):
Hey @Guilty-Spark-343, apologies just realised that there appears to be a unit test failing. If you're able to look into that, that would be great

Guilty-Spark-343 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/log/ConstellationLogFormatter.java:
```diff
@@ -0,0 +1,42 @@
+    public final String format(final LogRecord record) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append("[");
+        sb.append(Instant.now().toString().substring(0,22).replace("T"," "));
```
Yeah I agree LocalDateTime makes more sense given the logs are only used by the client

OrionsGuardian reviewed (changes requested):
It appears that Exception stack traces aren't being logged.
If you generate an exception via Experimental - > Developer -> Throw a RuntimeException
The log only shows one line without any detail:
`[2024-02-05 08:48:31.17Z] - SEVERE [global]: null`

It should be showing the full error stacktrace (as is shown in the Error Report):
```
java.lang.RuntimeException: Runtime exception successfully thrown at Mon Feb 05 08:48:26 ACDT 2024

au.gov.asd.tac.constellation.testing.RuntimeExceptionAction.actionPerformed(RuntimeExceptionAction.java:44)
org.openide.awt.AlwaysEnabledAction$1.run(AlwaysEnabledAction.java:174)
org.openide.util.actions.ActionInvoker$1.run(ActionInvoker.java:70)
org.openide.util.actions.ActionInvoker.doPerformAction(ActionInvoker.java:91)
org.openide.util.actions.ActionInvoker.invokeAction(ActionInvoker.java:74)
org.openide.awt.AlwaysEnabledAction.actionPerformed(AlwaysEnabledAction.java:177)
java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
java.desktop/javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2308)
java.desktop/javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:405)
java.desktop/javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:262)
java.desktop/javax.swing.AbstractButton.doClick(AbstractButton.java:369)
java.desktop/javax.swing.plaf.basic.BasicMenuItemUI.doClick(BasicMenuItemUI.java:1020)
java.desktop/javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(BasicMenuItemUI.java:1064)
java.desktop/java.awt.Component.processMouseEvent(Component.java:6632)
java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3342)
java.desktop/java.awt.Component.processEvent(Component.java:6397)
java.desktop/java.awt.Container.processEvent(Container.java:2263)
java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5008)
java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4547)
java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4488)
java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2307)
java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2772)
java.desktop/java.awt.Component.dispatchEvent(Component.java:4840)
java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:772)
java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
java.base/java.security.AccessController.doPrivileged(Native Method)
java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
java.desktop/java.awt.EventQueue$5.run(EventQueue.java:745)
java.desktop/java.awt.EventQueue$5.run(EventQueue.java:743)
java.base/java.security.AccessController.doPrivileged(Native Method)
java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:742)
org.netbeans.core.TimableEventQueue.dispatchEvent(TimableEventQueue.java:136)
java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```

Guilty-Spark-343 commented:
Is there a better way to view the results of the github unit tests? Locally it passes all of the tests but the github build is failing 2 tests in Utilities but I can't see what they are.

antares1470 commented:
> Is there a better way to view the results of the github unit tests? Locally it passes all of the tests but the github build is failing 2 tests in Utilities but I can't see what they are.

You can add the "verbose-logging" tag to your PR which will trigger the build and scan job in verbose mode (which has more detail on which tests failed).

That being said, it seems your most recent change appears to have fixed the unit test issue (its just now failing at the sonar part which all PRs are presently) so probably don't need to do it in this instance.

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/log/ConstellationLogFormatter.java:
```diff
@@ -44,9 +47,23 @@ public final String format(final LogRecord record) {
-        sb.append(formattedMessage);
         
-        sb.append(SeparatorConstants.NEWLINE);
+        Throwable error = record.getThrown();
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/log/ConstellationLogFormatter.java:
```diff
@@ -44,9 +47,23 @@ public final String format(final LogRecord record) {
+            sb.append(SeparatorConstants.NEWLINE);
+            sb.append(SeparatorConstants.NEWLINE);
+
+            String stackTrace = Arrays.stream(error.getStackTrace())
```
final

antares1470 commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/log/ConstellationLogFormatter.java:
```diff
@@ -44,9 +47,23 @@ public final String format(final LogRecord record) {
         
-        sb.append(SeparatorConstants.NEWLINE);
+        Throwable error = record.getThrown();
+        if (error != null && error.getMessage() != null && !error.getMessage().isBlank()) {
```
the latter 2 conditions can be replaced with `StringUtils.isNotBlank(error.getMessage())` which has a null check included

OrionsGuardian reviewed (changes requested):
When a new log message is identical to the previous log message, we should get a condensed line in the log instead of repeating the same log message:
`[08/02/2024 08:50:23] SEVERE [null]: Last record repeated again.`

This is also true for an exception stack trace message.
Rather than printing out the whole stacktrace again, we should get the condensed message.

You can test this with: Experimental ->  Developer -> Throw a RuntimeException from a plugin
(note it needs an active graph for the option to be available)

Guilty-Spark-343 commented:
Would it be if it was the exact same message or just partial match? The Runtime exception includes the datetime in the message so the error will be different each time e.g.  java.lang.RuntimeException: Runtime exception successfully thrown at Thu Feb 08 22:18:19 AEDT 2024

OrionsGuardian commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/log/LogPreferences.java:
```diff
@@ -29,6 +31,17 @@ public class LogPreferences {
+    private static final Logger LOGGER = Logger.getLogger(LogPreferences.class.getName());
+    
+    
+    static {
+        if( LOGGER.getUseParentHandlers() ){
+            final Handler[] parentHandlers = LOGGER.getParent().getHandlers();
+            for (final Handler handler : parentHandlers) {
+                handler.setFormatter(new ConstellationLogFormatter());
+            }
+        }
+    }
```
This code doesn't get called until the Data Access View is opened.
So, on a clean build with no Views loaded on startup, the ConstellationLogFormatter will not be used until after the Data Access View is opened.
Probably best to move this code to a class that is always accessed on startup.

OrionsGuardian commented:
Another example of how the current logger condenses repeated error messages:
```
[13/02/2024 14:39:25] FINE [au.gov.asd.tac.constellation.plugins.PluginExecution]: Quality Control View: Update State
[13/02/2024 14:39:25] FINE [null]: Last record repeated more than 10 times, further logs of this record are ignored until the log record changes.
[13/02/2024 14:39:40] FINE [null]: Last record repeated 589 times in total.
```

Guilty-Spark-343 commented:
Sorry about the delay on this one, I've moved the logging handler to Core Functionality startup script and added unit testing for the formatter.  Should be good to go if there are no other changes

OrionsGuardian commented:
Testing this in my dev environment I noticed that the changes are only being applied to the console output, and not being written into the messages.log file.
  
**messages.log** file:
```
java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
java.base/java.lang.Thread.run(Thread.java:1583)
[2024-04-11 08:50:00.02Z] - SEVERE [null]: Last record repeated more than 10 times, further logs of this record are ignored until the log record changes.
[2024-04-11 08:50:54.75Z] - SEVERE [null]: Last record repeated 25 times in total.
[2024-04-11 08:50:54.75Z] - FINE [au.gov.asd.tac.constellation.plugins.PluginExecution]: Throw a RuntimeException
[2024-04-11 08:50:54.75Z] - SEVERE [au.gov.asd.tac.constellation.graph.node.plugins.DefaultPluginEnvironment]: Unexpected exception caught in Throw a RuntimeException
```
  
**Console output:**
```
java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
java.base/java.lang.Thread.run(Thread.java:1583)
[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:56.69Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:57.21Z] - SEVERE [ScriptingView instance]: ^^Last record repeated again.[2024-04-11 08:49:57.21Z] - SEVERE [ScriptingView instance]: ^^Last record repeated more than 10 times, further logs of this record are ignored until the log record changes.[2024-04-11 08:50:50.26Z] - FINE [au.gov.asd.tac.constellation.plugins.PluginExecution]: Throw a RuntimeException
[2024-04-11 08:50:50.26Z] - SEVERE [au.gov.asd.tac.constellation.graph.node.plugins.DefaultPluginEnvironment]: Unexpected exception caught in Throw a RuntimeException
```

Note, I added the two caret symbols (^^) into my build, to confirm which code is being run for the console and the messages.log file. Also looks like the custom messages should contain a newline character at the end of the string.
Further note that it seems that some messages don't get written until a new/different log message is generated (via a different exception, or regular info log message).

For reference, this is the script I use for testing (within the Scripting View in Constellation):
```
print('\nTEST EXCEPTIONS SCRIPT STARTED')

from time import sleep
from java.util.logging import Logger
from java.util.logging import Level
import java.lang.Exception
import java.lang.Throwable

LOGGER = Logger.getLogger("ScriptingView instance");
LOGGER.setLevel(Level.ALL);

def generateFineError():
    try:
        raise java.lang.Exception("auto generated FINE error")
    except java.lang.Throwable, t:
        LOGGER.log(Level.FINE, "Re-routed exception ", t);

def generateInfoError():
    try:
        raise java.lang.Exception("auto generated INFO error")
    except java.lang.Throwable, t:
        LOGGER.log(Level.INFO, "Re-routed exception ", t);

def generateWarningError():
    try:
        raise java.lang.Exception("auto generated WARNING error")
    except java.lang.Throwable, t:
        LOGGER.log(Level.WARNING, "Re-routed exception ", t);

def generateSevereError():
    try:
        raise java.lang.Exception("auto generated SEVERE error")
    except java.lang.Throwable, t:
        LOGGER.log(Level.SEVERE, "Re-routed exception ", t);

for idx in range(0,25):
    #generateFineError()
    #generateInfoError()
    #generateWarningError()
    generateSevereError()
    sleep(0.05)  

print('\nTEST EXCEPTIONS SCRIPT COMPLETED')
```

OrionsGuardian commented on CoreUtilities/src/au/gov/asd/tac/constellation/utilities/log/ConstellationLogFormatter.java:
```diff
@@ -0,0 +1,91 @@
+            formattedMessage = logRecord.getMessage();
+        }
+
+        if(formattedMessage != null && logMessage != null && formattedMessage.equals(logMessage)){
+            if(logCount == 10){
+                sb.append("Last record repeated more than 10 times, further logs of this record are ignored until the log record changes.");
+            }else if(logCount > 10){
+                return ""; // Do not log messages until it changes
+            }else{
+                sb.append("Last record repeated again.");
+            }
+            logCount += 1;
+            return sb.toString();
+        }else {
+            logCount = 0;            
+        }
```
Code formatting: add spaces between keywords and brackets (can all be fixed with Alt-Shift-F in the IDE)

OrionsGuardian reviewed (changes requested):
Also address functionality issue described in earlier comment.

Auriga2 commented:
Experimental -> Developer -> Throw a RuntimeException shows a "null", and logs with the same time stamp rather than the repeating clause?
![image](https://github.com/user-attachments/assets/48e5735c-2854-4b3e-8b17-e06071ba4b20)

![image](https://github.com/user-attachments/assets/2d0a06f6-8a0c-425a-8807-c9ced0c735a5)

OrionsGuardian commented:
> Experimental -> Developer -> `Throw a RuntimeException` shows a "null" in the log.
> 
> Also repeat exceptions log multiple messages with the same time stamp rather than the repeating clause? ![image](https://private-user-images.githubusercontent.com/64826928/413656765-48e5735c-2854-4b3e-8b17-e06071ba4b20.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5MTUzNDMsIm5iZiI6MTczOTkxNTA0MywicGF0aCI6Ii82NDgyNjkyOC80MTM2NTY3NjUtNDhlNTczNWMtMjg1NC00YjNlLThiMTctZTA2MDcxYmE0YjIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE4VDIxNDQwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0NGYyODZlYTk0OGViMWI0ZGJlN2YxMTI1ZGJiODFhMzkzODVkYTgwZGY0ZmJhNGFkNWM5ZmY2ZTg0Y2RjMmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AdXr-TA_FDs6vY4skDuS3t0bcTNJg-t0TommZ6sGI20)
> 
> ![image](https://private-user-images.githubusercontent.com/64826928/413656787-2d0a06f6-8a0c-425a-8807-c9ced0c735a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk5MTUzNDMsIm5iZiI6MTczOTkxNTA0MywicGF0aCI6Ii82NDgyNjkyOC80MTM2NTY3ODctMmQwYTA2ZjYtOGEwYy00MjVhLTg4MDctYzljZWQwYzczNWE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjE4VDIxNDQwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVkYTFlZDRkYjlmZDg2OGY1YzJiZjQ1ZWJmYjYwY2FmOTgzOGI3NTAxMmE5YWUzOGY5NmMzNWM1NTMxODcxZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.DGw5uR7FgIulmHOSTeOH6VzkQFpF0LwM4DJ0uqasGOg)

I've managed to replicate the duplicate message which is occasionally shown in the Logs window in Constellation.
Looking at the messages.log file directly, it correctly shows that there is only 1 instance of the message.
So there is an issue with how the Show Logs functionality works in Constellation.
